File: simulation/src/nova_sim/scripts/moveit2_arm_executor.py
#!/usr/bin/env python3
"""
MoveIt2 IK + ros2_control executor for nova arms.

Input topic:
  /nova_arm_goal (std_msgs/String, JSON)

JSON schema:
{
  "arm_id": 0,
  "target_pose": {
    "frame_id": "base_link",
    "position": [0.35, -0.10, 0.25],
    "orientation": [0.0, 0.0, 0.0, 1.0]
  },
  "gripper": {
    "mode": "open" | "close" | "width",
    "value": 0.06
  },
  "execute_order": "pose_then_gripper" | "gripper_then_pose" | "sync"
}
"""

import json
from copy import deepcopy
from typing import Dict, Tuple

import rclpy
from rclpy.duration import Duration
from rclpy.node import Node

from geometry_msgs.msg import PoseStamped
from moveit_msgs.msg import Constraints, JointConstraint, PositionIKRequest, RobotState
from moveit_msgs.srv import GetPositionIK
from sensor_msgs.msg import JointState
from std_msgs.msg import Float64MultiArray, String


class MoveIt2ArmExecutor(Node):
    def __init__(self) -> None:
        super().__init__("moveit2_arm_executor")

        self.arm_groups = {
            0: "j1_arm",
            1: "j2_arm",
            2: "j3_arm",
            3: "j4_arm",
        }
        self.ee_links = {
            0: "J1_6",
            1: "J2_6",
            2: "J3_6",
            3: "J4_6",
        }
        self.arm_prefix = {
            0: "J1_",
            1: "J2_",
            2: "J3_",
            3: "J4_",
        }
        self.control_joint_order = [
            "J1_1_joint", "J1_2_joint", "J1_3_joint", "J1_4_joint", "J1_5_joint", "J1_6_joint",
            "J1_7_joint", "J1_8_joint",
            "J2_1_joint", "J2_2_joint", "J2_3_joint", "J2_4_joint", "J2_5_joint", "J2_6_joint",
            "J2_7_joint", "J2_8_joint",
            "J3_1_joint", "J3_2_joint", "J3_3_joint", "J3_4_joint", "J3_5_joint", "J3_6_joint",
            "J4_1_joint", "J4_2_joint", "J4_3_joint", "J4_4_joint", "J4_5_joint", "J4_6_joint",
        ]
        self.gripper_joints = {
            0: ("J1_7_joint", "J1_8_joint"),
            1: ("J2_7_joint", "J2_8_joint"),
        }
        # Tunable and explicit, so physical opening direction can be calibrated quickly.
        self.gripper_open = {
            0: (0.08, 0.08),
            1: (0.08, 0.08),
        }
        self.gripper_close = {
            0: (0.0, 0.0),
            1: (0.0, 0.0),
        }

        self.current_joint_map: Dict[str, float] = {}
        self.last_joint_state_time = self.get_clock().now()

        self.ik_client = self.create_client(GetPositionIK, "/compute_ik")
        # arm_controller is JointGroupPositionController, command topic is /arm_controller/commands
        self.command_pub = self.create_publisher(Float64MultiArray, "/arm_controller/commands", 10)
        self.joint_state_sub = self.create_subscription(JointState, "/joint_states", self.on_joint_state, 50)
        self.goal_sub = self.create_subscription(String, "/nova_arm_goal", self.on_goal, 10)

        self.get_logger().info("Waiting for /compute_ik service...")
        self.ik_client.wait_for_service()
        self.get_logger().info("MoveIt2 arm executor ready. Subscribe /nova_arm_goal")

    def on_joint_state(self, msg: JointState) -> None:
        for i, name in enumerate(msg.name):
            if i < len(msg.position):
                self.current_joint_map[name] = float(msg.position[i])
        self.last_joint_state_time = self.get_clock().now()

    def on_goal(self, msg: String) -> None:
        try:
            goal = json.loads(msg.data)
        except json.JSONDecodeError as exc:
            self.get_logger().error(f"Invalid JSON on /nova_arm_goal: {exc}")
            return

        try:
            arm_id = int(goal["arm_id"])
            pose_cfg = goal["target_pose"]
            frame_id = pose_cfg.get("frame_id", "base_link")
            pos = pose_cfg["position"]
            quat = pose_cfg["orientation"]
        except Exception as exc:  # noqa: BLE001
            self.get_logger().error(f"Bad goal payload: {exc}")
            return

        if arm_id not in self.arm_groups:
            self.get_logger().error(f"Unsupported arm_id={arm_id}, expect 0/1/2/3")
            return

        if not self.current_joint_map:
            self.get_logger().error("No /joint_states received yet, cannot seed IK.")
            return

        pose = PoseStamped()
        pose.header.frame_id = frame_id
        pose.header.stamp = self.get_clock().now().to_msg()
        pose.pose.position.x = float(pos[0])
        pose.pose.position.y = float(pos[1])
        pose.pose.position.z = float(pos[2])
        pose.pose.orientation.x = float(quat[0])
        pose.pose.orientation.y = float(quat[1])
        pose.pose.orientation.z = float(quat[2])
        pose.pose.orientation.w = float(quat[3])

        target_arm_joints = self.compute_ik(arm_id, pose)
        if target_arm_joints is None:
            self.get_logger().error(f"IK failed for arm_id={arm_id}")
            return

        order = goal.get("execute_order", "pose_then_gripper")
        gripper_cfg = goal.get("gripper")

        if order == "gripper_then_pose":
            if gripper_cfg is not None:
                self.execute_gripper(arm_id, gripper_cfg, time_s=0.8)
            self.execute_arm(arm_id, target_arm_joints, time_s=2.0)
        elif order == "sync":
            self.execute_arm_with_optional_gripper(arm_id, target_arm_joints, gripper_cfg, time_s=2.0)
        else:
            self.execute_arm(arm_id, target_arm_joints, time_s=2.0)
            if gripper_cfg is not None:
                self.execute_gripper(arm_id, gripper_cfg, time_s=0.8)

    def compute_ik(self, arm_id: int, pose: PoseStamped) -> Dict[str, float] | None:
        req = GetPositionIK.Request()
        ikr = PositionIKRequest()
        ikr.group_name = self.arm_groups[arm_id]
        ikr.ik_link_name = self.ee_links[arm_id]
        ikr.pose_stamped = pose
        ikr.timeout = Duration(seconds=0.5).to_msg()
        ikr.avoid_collisions = True

        state = RobotState()
        seed = JointState()
        seed.name = list(self.current_joint_map.keys())
        seed.position = [self.current_joint_map[n] for n in seed.name]
        state.joint_state = seed
        ikr.robot_state = state

        # Keep non-target arms near current posture to stabilize IK on shared model.
        constraints = Constraints()
        target_prefix = self.arm_prefix[arm_id]
        for name, pos in self.current_joint_map.items():
            if name.endswith("_joint") and not name.startswith(target_prefix):
                jc = JointConstraint()
                jc.joint_name = name
                jc.position = float(pos)
                jc.tolerance_above = 0.02
                jc.tolerance_below = 0.02
                jc.weight = 0.2
                constraints.joint_constraints.append(jc)
        ikr.constraints = constraints

        req.ik_request = ikr
        future = self.ik_client.call_async(req)
        rclpy.spin_until_future_complete(self, future, timeout_sec=1.0)
        if not future.done() or future.result() is None:
            self.get_logger().error("No IK response from /compute_ik")
            return None

        res = future.result()
        if res.error_code.val != 1:
            self.get_logger().error(f"IK failed, MoveIt code={res.error_code.val}")
            return None

        out = {}
        for i, name in enumerate(res.solution.joint_state.name):
            if i < len(res.solution.joint_state.position):
                out[name] = float(res.solution.joint_state.position[i])
        return out

    def resolve_gripper_targets(self, arm_id: int, cfg: dict) -> Tuple[str, str, float, float] | None:
        if arm_id not in self.gripper_joints:
            self.get_logger().warn(f"arm_id={arm_id} has no gripper, skip gripper command")
            return None
        j0, j1 = self.gripper_joints[arm_id]
        mode = str(cfg.get("mode", "open")).lower()
        if mode == "open":
            p0, p1 = self.gripper_open[arm_id]
        elif mode == "close":
            p0, p1 = self.gripper_close[arm_id]
        elif mode == "width":
            width = float(cfg.get("value", 0.06))
            width = max(0.0, min(0.1, width))
            ratio = width / 0.1
            o0, o1 = self.gripper_open[arm_id]
            c0, c1 = self.gripper_close[arm_id]
            p0 = c0 + (o0 - c0) * ratio
            p1 = c1 + (o1 - c1) * ratio
        else:
            self.get_logger().error(f"Unsupported gripper mode={mode}")
            return None
        return j0, j1, float(p0), float(p1)

    def execute_arm(self, arm_id: int, ik_solution: Dict[str, float], time_s: float) -> None:
        cmd = deepcopy(self.current_joint_map)
        prefix = self.arm_prefix[arm_id]
        for name, pos in ik_solution.items():
            if name.startswith(prefix) and name in cmd:
                cmd[name] = pos
        self.publish_command(cmd)
        self.get_logger().info(f"Arm {arm_id} command sent.")

    def execute_gripper(self, arm_id: int, gripper_cfg: dict, time_s: float) -> None:
        resolved = self.resolve_gripper_targets(arm_id, gripper_cfg)
        if resolved is None:
            return
        j0, j1, p0, p1 = resolved
        cmd = deepcopy(self.current_joint_map)
        cmd[j0] = p0
        cmd[j1] = p1
        self.publish_command(cmd)
        self.get_logger().info(f"Arm {arm_id} gripper command sent: {j0}={p0:.4f}, {j1}={p1:.4f}")

    def execute_arm_with_optional_gripper(
        self, arm_id: int, ik_solution: Dict[str, float], gripper_cfg: dict | None, time_s: float
    ) -> None:
        cmd = deepcopy(self.current_joint_map)
        prefix = self.arm_prefix[arm_id]
        for name, pos in ik_solution.items():
            if name.startswith(prefix) and name in cmd:
                cmd[name] = pos

        if gripper_cfg is not None:
            resolved = self.resolve_gripper_targets(arm_id, gripper_cfg)
            if resolved is not None:
                j0, j1, p0, p1 = resolved
                cmd[j0] = p0
                cmd[j1] = p1
        self.publish_command(cmd)
        self.get_logger().info(f"Arm {arm_id} synced arm+gripper trajectory sent.")

    def publish_command(self, command_map: Dict[str, float]) -> None:
        msg = Float64MultiArray()
        msg.data = [float(command_map.get(j, 0.0)) for j in self.control_joint_order]
        self.command_pub.publish(msg)


def main(args=None) -> None:
    rclpy.init(args=args)
    node = MoveIt2ArmExecutor()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()

File: simulation/src/nova_sim_mujoco/src/moveit2_arm_executor.cpp
// MoveIt2ArmExecutorCpp：GetPositionIK 请求构造、关节顺序与夹爪开合阈值。
#include "nova_sim/moveit2_arm_executor.hpp"

#include <chrono>
#include <cmath>

#include "moveit_msgs/msg/constraints.hpp"
#include "moveit_msgs/msg/joint_constraint.hpp"
#include "moveit_msgs/msg/position_ik_request.hpp"
#include "moveit_msgs/msg/robot_state.hpp"

#include "calib_sim_mujoco/msg/arm_pose.hpp"

using namespace std::chrono_literals;

MoveIt2ArmExecutorCpp::MoveIt2ArmExecutorCpp()
: Node("moveit2_arm_executor_cpp"), arm_id_(0)
{
  // 与当前 nova_moveit_config 中 SRDF 的 group / 末端 link 命名保持一致。
  arm_groups_ = {{0, "l_arm"}, {1, "r_arm"}, {2, "j3_arm"}, {3, "j4_arm"}};
  ee_links_ = {{0, "J1_6"}, {1, "J2_6"}, {2, "J3_6"}, {3, "J4_6"}};
  arm_prefix_ = {{0, "J1_"}, {1, "J2_"}, {2, "J3_"}, {3, "J4_"}};

  control_joint_order_ = {
    "J1_1_joint", "J1_2_joint", "J1_3_joint", "J1_4_joint", "J1_5_joint", "J1_6_joint", "J1_7_joint",
    "J1_8_joint", "J2_1_joint", "J2_2_joint", "J2_3_joint", "J2_4_joint", "J2_5_joint", "J2_6_joint",
    "J2_7_joint", "J2_8_joint", "J3_1_joint", "J3_2_joint", "J3_3_joint", "J3_4_joint", "J3_5_joint",
    "J3_6_joint", "J4_1_joint", "J4_2_joint", "J4_3_joint", "J4_4_joint", "J4_5_joint", "J4_6_joint"};

  gripper_joints_ = {{0, {"J1_7_joint", "J1_8_joint"}}, {1, {"J2_7_joint", "J2_8_joint"}}};
  gripper_open_ = {{0, {0.02, -0.02}}, {1, {0.02, -0.02}}};
  gripper_close_ = {{0, {-0.04, 0.04}}, {1, {-0.04, 0.04}}};

  ik_client_ = this->create_client<moveit_msgs::srv::GetPositionIK>("/compute_ik");
  command_pub_ = this->create_publisher<std_msgs::msg::Float64MultiArray>("/arm_controller/commands", 10);
  pose_log_pub_ = this->create_publisher<std_msgs::msg::String>("/nova_pose_log", 20);

  joint_state_sub_ = this->create_subscription<sensor_msgs::msg::JointState>(
    "/joint_states", 50, std::bind(&MoveIt2ArmExecutorCpp::on_joint_state, this, std::placeholders::_1));
  arm_id_sub_ = this->create_subscription<std_msgs::msg::Int32>(
    "/nova_arm_id", 10, std::bind(&MoveIt2ArmExecutorCpp::on_arm_id, this, std::placeholders::_1));
  target_arm_pose_sub_ = this->create_subscription<calib_sim_mujoco::msg::ArmPose>(
    "/nova_target_arm_pose", 10, std::bind(&MoveIt2ArmExecutorCpp::on_target_arm_pose, this, std::placeholders::_1));
  gripper_sub_ = this->create_subscription<std_msgs::msg::String>(
    "/nova_gripper_goal", 10, std::bind(&MoveIt2ArmExecutorCpp::on_gripper_goal, this, std::placeholders::_1));

  RCLCPP_INFO(this->get_logger(), "moveit2_arm_executor_cpp ready.");
  RCLCPP_INFO(
    this->get_logger(),
    "[nova_executor_v2] IK request mode: seeded by /joint_states, groups(l_arm/r_arm).");
  if (!ik_client_->wait_for_service(200ms)) {
    RCLCPP_WARN(
      this->get_logger(),
      "/compute_ik not ready now. Joint/gripper control still works; pose control waits for MoveIt.");
  }
  RCLCPP_INFO(
    this->get_logger(),
    "Pose goals: calib_sim_mujoco/ArmPose on /nova_target_arm_pose only "
    "(/nova_arm_id + /nova_target_pose 不再被此节点订阅，避免错序). + optional /nova_gripper_goal.");
}

void MoveIt2ArmExecutorCpp::on_joint_state(const sensor_msgs::msg::JointState::SharedPtr msg)
{
  for (size_t i = 0; i < msg->name.size() && i < msg->position.size(); ++i) {
    current_joint_map_[msg->name[i]] = msg->position[i];
  }
}

void MoveIt2ArmExecutorCpp::on_arm_id(const std_msgs::msg::Int32::SharedPtr msg)
{
  if (arm_groups_.count(msg->data) == 0) {
    RCLCPP_ERROR(
      this->get_logger(),
      "Invalid arm_id=%d for pose IK. Current MoveIt config supports arm_id 0/1/2/3.",
      msg->data);
    return;
  }
  arm_id_ = msg->data;
  RCLCPP_INFO(this->get_logger(), "Set active arm_id=%d", arm_id_);
}

void MoveIt2ArmExecutorCpp::on_target_arm_pose(const calib_sim_mujoco::msg::ArmPose::SharedPtr msg)
{
  process_pose_goal(msg->arm_id, msg->pose);
}

void MoveIt2ArmExecutorCpp::process_pose_goal(
  int arm_id, const geometry_msgs::msg::PoseStamped & pose)
{
  if (arm_groups_.count(arm_id) == 0) {
    RCLCPP_ERROR(
      this->get_logger(),
      "Invalid arm_id=%d for pose IK. Current MoveIt config supports arm_id 0/1/2/3.",
      arm_id);
    return;
  }

  if (!ik_client_->wait_for_service(100ms)) {
    RCLCPP_ERROR(this->get_logger(), "Pose goal rejected: /compute_ik service is unavailable.");
    std_msgs::msg::String log_msg;
    log_msg.data = "[ERROR] Pose goal rejected: /compute_ik service is unavailable.";
    pose_log_pub_->publish(log_msg);
    return;
  }

  if (current_joint_map_.empty()) {
    RCLCPP_ERROR(this->get_logger(), "No /joint_states yet, ignore pose goal.");
    std_msgs::msg::String log_msg;
    log_msg.data = "[ERROR] No /joint_states yet, ignore pose goal.";
    pose_log_pub_->publish(log_msg);
    return;
  }

  auto ik_req = std::make_shared<moveit_msgs::srv::GetPositionIK::Request>();
  moveit_msgs::msg::PositionIKRequest ikr;
  ikr.group_name = arm_groups_[arm_id];
  ikr.ik_link_name = ee_links_[arm_id];
  ikr.pose_stamped = pose;
  const double qx = ikr.pose_stamped.pose.orientation.x;
  const double qy = ikr.pose_stamped.pose.orientation.y;
  const double qz = ikr.pose_stamped.pose.orientation.z;
  const double qw = ikr.pose_stamped.pose.orientation.w;
  const double qnorm = std::sqrt(qx * qx + qy * qy + qz * qz + qw * qw);
  if (qnorm < 1e-8) {
    RCLCPP_ERROR(this->get_logger(), "Pose goal rejected: invalid quaternion norm=%g", qnorm);
    std_msgs::msg::String log_msg;
    log_msg.data = "[ERROR] Pose goal rejected: invalid quaternion norm.";
    pose_log_pub_->publish(log_msg);
    return;
  }
  ikr.pose_stamped.pose.orientation.x = qx / qnorm;
  ikr.pose_stamped.pose.orientation.y = qy / qnorm;
  ikr.pose_stamped.pose.orientation.z = qz / qnorm;
  ikr.pose_stamped.pose.orientation.w = qw / qnorm;
  ikr.timeout.sec = 1;
  ikr.timeout.nanosec = 0;
  // For current bring-up phase, prioritize solvability first.
  ikr.avoid_collisions = false;

  // Seed IK with current arm joint values; avoid empty JointState in MoveIt conversions.
  sensor_msgs::msg::JointState seed;
  seed.name.reserve(current_joint_map_.size());
  seed.position.reserve(current_joint_map_.size());
  for (const auto & kv : current_joint_map_) {
    if (kv.first.rfind("J1_", 0) == 0 || kv.first.rfind("J2_", 0) == 0 ||
      kv.first.rfind("J3_", 0) == 0 || kv.first.rfind("J4_", 0) == 0)
    {
      seed.name.push_back(kv.first);
      seed.position.push_back(kv.second);
    }
  }
  if (seed.name.empty()) {
    RCLCPP_ERROR(this->get_logger(), "IK seed is empty, wait /joint_states update.");
    std_msgs::msg::String log_msg;
    log_msg.data = "[ERROR] IK seed is empty, wait /joint_states update.";
    pose_log_pub_->publish(log_msg);
    return;
  }
  ikr.robot_state.joint_state = seed;

  // Do not inject external joint constraints here.
  // xtrainer_moveit_config uses a different naming set, forcing unknown joints will fail IK request.
  ik_req->ik_request = ikr;
  RCLCPP_INFO(
    this->get_logger(),
    "[nova_executor_v2] send IK arm_id=%d group=%s link=%s frame=%s",
    arm_id,
    ikr.group_name.c_str(),
    ikr.ik_link_name.c_str(),
    ikr.pose_stamped.header.frame_id.c_str());

  const int request_arm_id = arm_id;
  const auto request_prefix = arm_prefix_[request_arm_id];
  ik_client_->async_send_request(
    ik_req,
    [this, request_arm_id, request_prefix](
      rclcpp::Client<moveit_msgs::srv::GetPositionIK>::SharedFuture future) {
      auto result = future.get();
      if (result->error_code.val != 1) {
        RCLCPP_ERROR(
          this->get_logger(), "IK failed for arm_id=%d, code=%d", request_arm_id, result->error_code.val);
        std_msgs::msg::String log_msg;
        log_msg.data = "[ERROR] IK failed for arm_id=" + std::to_string(request_arm_id) +
          ", code=" + std::to_string(result->error_code.val);
        pose_log_pub_->publish(log_msg);
        return;
      }

      std::unordered_map<std::string, double> command_map = current_joint_map_;
      const auto & names = result->solution.joint_state.name;
      const auto & pos = result->solution.joint_state.position;
      for (size_t i = 0; i < names.size() && i < pos.size(); ++i) {
        std::string target_name = names[i];
        // Map MoveIt naming (R1-1/R2-3...) to controller naming (J1_1_joint/J2_3_joint...)
        if (target_name.rfind("R1-", 0) == 0 && target_name.size() > 3) {
          target_name = "J1_" + target_name.substr(3) + "_joint";
        } else if (target_name.rfind("R2-", 0) == 0 && target_name.size() > 3) {
          target_name = "J2_" + target_name.substr(3) + "_joint";
        }

        if (target_name.rfind(request_prefix, 0) == 0) {
          command_map[target_name] = pos[i];
        }
      }
      publish_command(command_map);
      arm_id_ = request_arm_id;
      RCLCPP_INFO(this->get_logger(), "Pose command sent for arm_id=%d", request_arm_id);
      std_msgs::msg::String log_msg;
      log_msg.data = "[INFO] Pose command sent for arm_id=" + std::to_string(request_arm_id);
      pose_log_pub_->publish(log_msg);
    });
}

void MoveIt2ArmExecutorCpp::on_gripper_goal(const std_msgs::msg::String::SharedPtr msg)
{
  if (gripper_joints_.count(arm_id_) == 0) {
    RCLCPP_WARN(this->get_logger(), "arm_id=%d has no gripper", arm_id_);
    return;
  }
  std::unordered_map<std::string, double> command_map = current_joint_map_;
  const auto joints = gripper_joints_[arm_id_];
  double p0 = 0.0;
  double p1 = 0.0;
  if (msg->data == "open") {
    p0 = gripper_open_[arm_id_].first;
    p1 = gripper_open_[arm_id_].second;
  } else if (msg->data == "close") {
    p0 = gripper_close_[arm_id_].first;
    p1 = gripper_close_[arm_id_].second;
  } else if (msg->data.rfind("width:", 0) == 0) {
    const auto value = std::stod(msg->data.substr(6));
    const auto width = std::max(0.0, std::min(0.06, value));
    const auto ratio = width / 0.06;
    p0 = gripper_close_[arm_id_].first +
      (gripper_open_[arm_id_].first - gripper_close_[arm_id_].first) * ratio;
    p1 = gripper_close_[arm_id_].second +
      (gripper_open_[arm_id_].second - gripper_close_[arm_id_].second) * ratio;
  } else {
    RCLCPP_ERROR(this->get_logger(), "Bad /nova_gripper_goal: use open|close|width:<m>");
    return;
  }
  command_map[joints.first] = p0;
  command_map[joints.second] = p1;
  publish_command(command_map);
  RCLCPP_INFO(this->get_logger(), "Gripper command sent for arm_id=%d", arm_id_);
}

void MoveIt2ArmExecutorCpp::publish_command(const std::unordered_map<std::string, double> & command_map)
{
  constexpr double kJointDeadband = 3e-4;
  bool has_meaningful_change = !has_last_command_;
  if (!has_meaningful_change) {
    for (const auto & joint : control_joint_order_) {
      const double new_val = resolve_command_value(joint, command_map);
      const auto it_old = last_published_command_map_.find(joint);
      const double old_val = (it_old == last_published_command_map_.end()) ? 0.0 : it_old->second;
      if (std::abs(new_val - old_val) > kJointDeadband) {
        has_meaningful_change = true;
        break;
      }
    }
  }

  if (!has_meaningful_change) {
    return;
  }

  std_msgs::msg::Float64MultiArray msg;
  msg.data.reserve(control_joint_order_.size());
  for (const auto & joint : control_joint_order_) {
    const double value = resolve_command_value(joint, command_map);
    msg.data.push_back(value);
    last_published_command_map_[joint] = value;
  }
  has_last_command_ = true;
  command_pub_->publish(msg);
}

double MoveIt2ArmExecutorCpp::resolve_command_value(
  const std::string & joint, const std::unordered_map<std::string, double> & command_map) const
{
  const auto it = command_map.find(joint);
  if (it != command_map.end()) {
    return it->second;
  }
  const auto it_last = last_published_command_map_.find(joint);
  if (it_last != last_published_command_map_.end()) {
    return it_last->second;
  }
  const auto it_cur = current_joint_map_.find(joint);
  if (it_cur != current_joint_map_.end()) {
    return it_cur->second;
  }
  return 0.0;
}
